### PR TITLE
Fix false positive with Generator when yield is in a "while" condition

### PR DESCRIPTION
**Pull request by dantleech**
Relates to: https://github.com/phpstan/phpstan/issues/3669

This PR fixes the above named issue as suggested there.

Not sure about the test however.

### CODE REVIEW

File: tests/PHPStan/Rules/Functions/data/returnTypes.php [hunk 1]
```diff
@@ -172,3 +172,11 @@ function returnVoidFromGenerator2(): \Generator
 	yield 1;
 	return 2;
 }
+
+/**
+ * @return int
+ */
+function returnFromYieldInCondition()
+{
+	return 'foobar';
```
Comment (dantleech) [inline tests/PHPStan/Rules/Functions/data/returnTypes.php:181]: should fail the test, but there is no additional message

File: tests/PHPStan/Rules/Missing/data/missing-return.php [hunk 2]
```diff
@@ -396,6 +396,12 @@ public function bodySpecifiedVoidTReturn3(): \Generator
 		return 2;
 	}
 
+	public function yieldInExpression(): \Generator
```
Comment (dantleech) [inline tests/PHPStan/Rules/Missing/data/missing-return.php:399]: The issue reported was detected by the MissingReturn rule. Not sure if there is a more suitable way to test (the test is that there is no error).

File: src/Analyser/NodeScopeResolver.php [hunk 3]
```diff
@@ -920,15 +924,23 @@ private function processStmtNode(
 				$finalScope = $scope;
 			}
 			if (!$alwaysTerminating) {
-				$finalScope = $this->processExprNode($stmt->cond, $bodyScope, $nodeCallback, ExpressionContext::createDeep())->getFalseyScope();
+				$stmtResult = $this->processExprNode($stmt->cond, $bodyScope, $nodeCallback, ExpressionContext::createDeep());
```
Comment (dantleech) [inline src/Analyser/NodeScopeResolver.php:927]: rename to condResult

File: src/Analyser/NodeScopeResolver.php [hunk 4]
```diff
@@ -955,8 +967,10 @@ private function processStmtNode(
 					$bodyScope = $bodyScope->mergeWith($continueExitPoint->getScope());
 				}
 				foreach ($stmt->loop as $loopExpr) {
-					$bodyScope = $this->processExprNode($loopExpr, $bodyScope, static function (): void {
-					}, ExpressionContext::createTopLevel())->getScope();
+					$exprResult = $this->processExprNode($loopExpr, $bodyScope, static function (): void {
+					}, ExpressionContext::createTopLevel());
+					$bodyScope = $exprResult->getScope();
+					$hasYield = $hasYield || $exprResult->hasYield();
```
Comment (dantleech) [inline src/Analyser/NodeScopeResolver.php:973]: not sure why this is a special case - i guess we could also get the `ExpressionResult` before this loop on `$stmt->loop`

File: tests/e2e/magic-setter/test.php [hunk 5]
```diff
@@ -1,4 +1,6 @@
-<?php // lint >= 7.4
+<?php declare(strict_types = 1);
+
+// lint >= 7.4
```
Comment (dantleech) [inline tests/e2e/magic-setter/test.php:3]: this was the cs fixer

**Comment by dantleech** [comment]
@ondrejmirtes updated for `for`, `do` and `foreach`

**Comment by ondrejmirtes** [comment]
Looks like a simple if statement can't handle it either: https://phpstan.org/r/5dafc496-79af-4719-bcf2-41fa369da313

**Comment by dantleech** [comment]
indeed, added `if` and `switch`

File: tests/e2e/magic-setter/test.php (hunk 5 above)
Comment (ondrejmirtes) [inline tests/e2e/magic-setter/test.php:3]: Nope, we don't want this change, php-parallel-lint doesn't understand this

Comment (dantleech) [reply tests/e2e/magic-setter/test.php]: Reverted

**Comment by ondrejmirtes** [comment]
Thank you!

**Pull request merged**